src/org/rascalmpl/library/util/Eval.java

```diff
@@ -48,169 +61,278 @@
 
-		Result<IValue> result = null;
-		Environment old = evaluator.getCurrentEnvt();
-		ModuleEnvironment env = getUniqueModuleEnvironment();
+	/**
+	 * A RascalRuntime is an evaluator that configures itself from a PathConfig instance,
+	 * and then offers the `eval` method, and the `reset` method to interact it.
+	 */
+	private static class RascalRuntime {
+		private final Evaluator eval;
+		private int duration = -1;
 		
-		try {
-			timer.start();
-			
-			evaluator.setCurrentEnvt(env);
-			if(!timer.hasExpired() && commands.length() > 0){
-				for(IValue command : commands){
-					ISourceLocation commandLocation = eval.getValueFactory().sourceLocation("eval", "", "/","command=" + ((IString)command).getValue(), null);
-					result = evaluator.evalMore(null, ((IString) command).getValue(), commandLocation);
+		public RascalRuntime(PathConfig pcfg, InputStream input, OutputStream stderr, OutputStream stdout, IDEServices services) throws IOException, URISyntaxException{
+			GlobalEnvironment heap = new GlobalEnvironment();
+			ModuleEnvironment root = heap.addModule(new ModuleEnvironment(ModuleEnvironment.SHELL_MODULE, heap));
+			IValueFactory vf = ValueFactoryFactory.getValueFactory();
+			this.eval = new Evaluator(vf, input, stderr, stdout, root, heap, services);
+
+			eval.addRascalSearchPathContributor(StandardLibraryContributor.getInstance());
+			eval.setMonitor(services);        
+			eval.getConfiguration().setRascalJavaClassPathProperty(javaCompilerPathAsString(pcfg.getJavaCompilerPath()));
+			eval.setMonitor(services);
+
+			if (!pcfg.getSrcs().isEmpty()) {
+				ISourceLocation projectRoot = inferProjectRoot((ISourceLocation) pcfg.getSrcs().get(0));
+				String projectName = new RascalManifest().getProjectName(projectRoot);
+				URIResolverRegistry reg = URIResolverRegistry.getInstance();
+				reg.registerLogical(new ProjectURIResolver(projectRoot, projectName));
+				reg.registerLogical(new TargetURIResolver(projectRoot, projectName));
+			}
+
+			for (IValue path : pcfg.getSrcs()) {
+				eval.addRascalSearchPath((ISourceLocation) path); 
+			}
+
+			for (IValue path : pcfg.getLibs()) {
+				eval.addRascalSearchPath((ISourceLocation) path);
+			}
+
+			ClassLoader cl = new SourceLocationClassLoader(pcfg.getClassloaders(), ShellEvaluatorFactory.class.getClassLoader());
+			eval.addClassLoader(cl);
+		}
+
+		public IValue staticTypeOf(String line) {
+			var result = eval.eval(null, line, IRascalValueFactory.getInstance().sourceLocation(URIUtil.assumeCorrect("eval", "", "", "command=" + line)));
+			var type = result.getStaticType();
+			return new TypeReifier(eval.getValueFactory()).typeToValue(type, new TypeStore(), eval.getValueFactory().map());
+		}
+
+		public void setTimeout(int duration) {
+			this.duration = duration;
+		}
+		
+		public int getTimeoutDuration() {
+			return duration;
+		}
+
+		private static ISourceLocation inferProjectRoot(ISourceLocation member) {
```

Comment:
At some point we're going to have to cleanup all the copies of this logic.

Reply:
agreed